File: app/src/main/java/io/github/freewebmovement/igniter/persistence/data/AppInfo.java
package io.github.freewebmovement.igniter.persistence.data;

import android.graphics.drawable.Drawable;

import androidx.annotation.NonNull;

public class AppInfo implements Cloneable {
    private String appName;
    private Drawable icon;
    private String packageName;
    private boolean enabled;

    public String getAppName() {
        return appName;
    }

    public void setAppName(String appName) {
        this.appName = appName;
    }

    public Drawable getIcon() {
        return icon;
    }

    public void setIcon(Drawable icon) {
        this.icon = icon;
    }

    public String getPackageName() {
        return packageName;
    }

    public void setPackageName(String packageName) {
        this.packageName = packageName;
    }

    public boolean getEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    @NonNull
    @Override
    protected Object clone() throws CloneNotSupportedException {
        AppInfo appInfo = (AppInfo) super.clone();
        appInfo.appName = appName;
        appInfo.icon = icon;
        appInfo.packageName = packageName;
        appInfo.enabled = enabled;
        return appInfo;
    }
}

File: app/src/main/java/io/github/freewebmovement/igniter/persistence/data/AppInfo.kt
package io.github.freewebmovement.igniter.persistence.data

import android.graphics.drawable.Drawable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

class AppInfo : Cloneable {
    var appName: String? = null
    var icon: Drawable? = null
    var packageName: String? = null
    var isSystemApp: Boolean = false

    /**
     * Backed by Compose state so the proxy-apps list reacts immediately when a
     * row's switch is toggled, even though the AppInfo object itself is reused.
     */
    var enabled: Boolean by mutableStateOf(false)

    @Throws(CloneNotSupportedException::class)
    override fun clone(): Any {
        val appInfo = super.clone() as AppInfo
        appInfo.appName = appName
        appInfo.icon = icon
        appInfo.packageName = packageName
        appInfo.enabled = enabled
        appInfo.isSystemApp = isSystemApp
        return appInfo
    }
}
